### PR TITLE
feat: create Roles and Rolebindings from SpaceBindings

### PR DESCRIPTION
**Pull request by xcoulon**
Uses the SpaceRoles defined in the NSTemplateTier associated
with the Space to create the Role and RoleBinding for the user
in the target Namespace(s).

also, include username in rolebinding names

see also:
- https://github.com/codeready-toolchain/api/pull/304
- https://github.com/codeready-toolchain/toolchain-common/pull/255
- https://github.com/codeready-toolchain/toolchain-e2e/pull/518

Fixes CRT-1351

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: controllers/space/space_controller.go [hunk 1]
```diff
@@ -309,6 +323,33 @@ func NewNSTemplateSetSpec(tmplTier *toolchainv1alpha1.NSTemplateTier) toolchainv
...
+			sr := tmplTier.Spec.SpaceRoles[r]
+			usernames := []string{}
+			for _, b := range bindings {
+				if b.Spec.SpaceRole == r {
+					usernames = append(usernames, b.Spec.MasterUserRecord)
+				}
+			}
+			// no need to add an entry in space roles if there is no associated user
+			if len(usernames) > 0 {
+				s.SpaceRoles = append(s.SpaceRoles, toolchainv1alpha1.NSTemplateSetSpaceRole{
```
Comment (MatousJobanek) [inline controllers/space/space_controller.go:345]: shouldn't we sort also the usernames so the order is always the same? 
```suggestion
				sort.Strings(usernames)
				s.SpaceRoles = append(s.SpaceRoles, toolchainv1alpha1.NSTemplateSetSpaceRole{
```

File: controllers/space/space_controller_test.go [hunk 2]
```diff
@@ -1029,6 +1031,127 @@ func TestUpdateSpaceTier(t *testing.T) {
...
+		// Space should be in "updating" state while the
+		spacetest.AssertThatSpace(t, test.HostOperatorNs, s.Name, hostClient).
+			HasConditions(spacetest.Updating())
+		// NSTemplateSet should have an spaceRoles entry for the `mur`
+		nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, nstmplSet.Name, member1Client).
+			HasSpaceRoleForUser("basic-admin-123456new", mur.Name). // entry added for user
+			HasConditions(nstemplatetsettest.Provisioned())         // not changed by the SpaceController, but will be by the NSTemplateSetController
+	})
+
+	t.Run("remove user with admin role", func(t *testing.T) {
```
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:1080]: could we modify the test case so in the `// given` part it would contain two space roles for one or two usernames? Then verify the removal of one SpaceBinding so in the `// then` part the `NSTemplateSet` would contain one SpaceRole/username less.

File: controllers/space/space_controller_test.go [hunk 3]
```diff
@@ -1029,6 +1031,127 @@ func TestUpdateSpaceTier(t *testing.T) {
...
+		spacetest.AssertThatSpace(t, test.HostOperatorNs, s.Name, hostClient).
+			HasConditions(spacetest.Updating())
+		// NSTemplateSet should have an spaceRoles entry for the `mur`
+		nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, nstmplSet.Name, member1Client).
+			HasNoSpaceRoleForUser("basic-viewer-123456new", mur.Name). // no change here
+			HasNoSpaceRoleForUser("basic-admin-123456new", mur.Name).  // entry removed for user
+			HasConditions(nstemplatetsettest.Provisioned())            // not changed by the SpaceController, but will be by the NSTemplateSetController
+	})
+
+	t.Run("update user from viewer to admin role", func(t *testing.T) {
```
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:1117]: analogically the same here - could we add more SpaceRoles/SpaceBindings as "noise" and verify that the controller updated only one?

File: deploy/templates/nstemplatetiers/test/ns_dev.yaml [hunk 4]
```diff
@@ -13,64 +13,6 @@ objects:
...
-    namespace: ${USERNAME}-dev
-  spec:
-    limits:
-    - type: "Container"
-      default:
-        cpu: 1000m
-        memory: 750Mi
-      defaultRequest:
-        cpu: 10m
-        memory: 64Mi
```
Comment (MatousJobanek) [inline deploy/templates/nstemplatetiers/test/ns_dev.yaml:73]: why did you remove this LimitRange?

File: deploy/templates/nstemplatetiers/test/ns_stage.yaml [hunk 5]
```diff
@@ -13,64 +13,6 @@ objects:
...
-    namespace: ${USERNAME}-stage
-  spec:
-    limits:
-    - type: "Container"
-      default:
-        cpu: 1000m
-        memory: 750Mi
-      defaultRequest:
-        cpu: 10m
-        memory: 64Mi
```
Comment (MatousJobanek) [inline deploy/templates/nstemplatetiers/test/ns_stage.yaml:73]: The same here

File: controllers/space/nstemplateset_spaces_mapper.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+			}}
+		}
+		// the obj was not a NSTemplateSet
+		mapperLog.Error(errors.New("not a NSTemplateSet"),
+			"MapNSTemplateSetToSpace attempted to map an object that was not a NSTemplateSet",
+			"obj", obj)
+		return []reconcile.Request{}
+	}
+
+}
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/nstemplateset_spaces_mapper.go:31]: hmm, indeed, it's not used (anymore) 🤔 Maybe a left-over from a previous version :/ 
I'll remove it

[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/nstemplateset_spaces_mapper.go]: removed in b60f306

File: controllers/space/nstemplateset_spaces_mapper.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	return func(obj client.Object) []reconcile.Request {
+		if sb, ok := obj.(*toolchainv1alpha1.SpaceBinding); ok {
+			if spaceName, found := sb.Labels[toolchainv1alpha1.SpaceBindingSpaceLabelKey]; found {
+				return []reconcile.Request{{
+					NamespacedName: types.NamespacedName{
+						Namespace: sb.Namespace,
+						Name:      spaceName,
+					},
+				}}
+			}
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/nstemplateset_spaces_mapper.go]: sure, let me try that

[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/nstemplateset_spaces_mapper.go]: done in 71c95d4

File: controllers/space/space_controller.go [hunk 8]
```diff
@@ -47,7 +48,9 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, memberClusters map[strin
 		For(&toolchainv1alpha1.Space{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
 		Watches(&source.Kind{Type: &toolchainv1alpha1.NSTemplateTier{}},
 			handler.EnqueueRequestsFromMapFunc(MapNSTemplateTierToSpaces(r.Namespace, r.Client)),
-			builder.WithPredicates(predicate.GenerationChangedPredicate{}))
+			builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(&source.Kind{Type: &toolchainv1alpha1.SpaceBinding{}},
+			handler.EnqueueRequestsFromMapFunc(MapSpaceBindingToSpace()))
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/space_controller.go]: done in 71c95d4

File: controllers/space/space_controller.go [hunk 9]
```diff
@@ -191,6 +194,16 @@ func (r *Reconciler) ensureNSTemplateSet(logger logr.Logger, space *toolchainv1a
 	}, tmplTier); err != nil {
 		return norequeue, r.setStatusProvisioningFailed(logger, space, err)
 	}
+	spacebindings := toolchainv1alpha1.SpaceBindingList{}
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/space_controller.go]: done in b3b60b0

File: deploy/templates/nstemplatetiers/test/ns_dev.yaml (hunk 4 above)
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_dev.yaml:73]: definitely NOT intended! I'll restore that

Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_dev.yaml:73]: restored in 6947622

File: deploy/templates/nstemplatetiers/test/ns_stage.yaml (hunk 5 above)
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_stage.yaml:73]: restored in 6947622

File: controllers/space/space_controller.go (hunk 1 above)
Comment (xcoulon) [reply controllers/space/space_controller.go:346]: good point!

Comment (xcoulon) [reply controllers/space/space_controller.go:347]: done in 144bef3

File: controllers/space/space_controller_test.go (hunk 2 above)
Comment (xcoulon) [reply controllers/space/space_controller_test.go:1093]: done in 3c4927f, with 2 extra space bindings to verify that the existing entries remained unchanged.

File: controllers/space/space_controller_test.go (hunk 3 above)
Comment (xcoulon) [reply controllers/space/space_controller_test.go:1143]: done in 9d23775

File: controllers/space/space_controller_test.go [hunk 10]
```diff
@@ -1029,6 +1031,127 @@ func TestUpdateSpaceTier(t *testing.T) {
...
+
+		// then
+		require.NoError(t, err)
+		assert.True(t, res.Requeue) // expect a requeue since the NSTemplateSet was updated
+		// Space should be in "updating" state while the
+		spacetest.AssertThatSpace(t, test.HostOperatorNs, s.Name, hostClient).
+			HasConditions(spacetest.Updating())
+		// NSTemplateSet should have an spaceRoles entry for the `mur`
+		nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, nstmplSet.Name, member1Client).
+			HasSpaceRoleForUser("basic-admin-123456new", mur.Name). // entry added for user
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/space_controller_test.go]: done in https://github.com/codeready-toolchain/host-operator/pull/616/commits/3c4927f06d4c760d1b403b5637862c5bea061963

File: controllers/space/space_controller_test.go [hunk 11]
```diff
@@ -1029,6 +1031,165 @@ func TestUpdateSpaceTier(t *testing.T) {
...
+		assert.True(t, res.Requeue) // expect a requeue since the NSTemplateSet was updated
+		// Space should be in "updating" state while the
+		spacetest.AssertThatSpace(t, test.HostOperatorNs, s.Name, hostClient).
+			HasConditions(spacetest.Updating())
+		// NSTemplateSet should have an spaceRoles entry for the `mur`
+		nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, nstmplSet.Name, member1Client).
+			HasSpaceRoleForUser("basic-admin-123456new", johnMUR.Name).    // entry added for user as admin
+			HasNoSpaceRoleForUser("basic-viewer-123456new", johnMUR.Name). // but not added as viewer
+			HasSpaceRoleForUser("basic-admin-123456new", adminMUR.Name).   // unchanged
+			HasSpaceRoleForUser("basic-viewer-123456new", viewerMUR.Name). // unchanged
```
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:1091]: Personally, I would rather use something like this:

```go
nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, nstmplSet.Name, member1Client).
	HasSpaceRoles(
		SpaceRole("basic-admin-123456new", johnMUR.Name, adminMUR.Name),
		SpaceRole("basic-viewer-123456new", viewerMUR.Name)).
```
In that way, the `HasSpaceRoles` method could easily verify that there is exactly this set of SpaceRoles defined (not more) and at the same time the `SpaceRole` function would verify that it has the exact set of usernames (not more either). So, compared to your version it would verify also the exact count/set of the entries plus it would be shorter. 
Just to be clear, this is a proposal - not a request that would block a PR ;-).

File: controllers/space/space_controller_test.go [hunk 12]
```diff
@@ -1029,6 +1031,165 @@ func TestUpdateSpaceTier(t *testing.T) {
...
+	t.Run("remove user with admin role", func(t *testing.T) {
+		// given a MUR, a Space and its NSTemplateSet resource...
+		s := spacetest.NewSpace("oddity",
+			spacetest.WithTierName(basicTier.Name),
+			spacetest.WithSpecTargetCluster("member-1"),
+			spacetest.WithStatusTargetCluster("member-1"), // already provisioned on a target cluster
+			spacetest.WithFinalizer())
+		adminMUR := murtest.NewMasterUserRecord(t, "jack")
+		viewerMUR := murtest.NewMasterUserRecord(t, "jeff")
+		johnMUR := murtest.NewMasterUserRecord(t, "john")
```
Comment (MatousJobanek) [inline controllers/space/space_controller_test.go:1104]: it looks like that this set of MURs (plus maybe some additional resources like the `Space` as well as the `SpaceBinding`) are used in all newly added tests - you could instantiate them only once in the parent test and just reuse them. This would make the test much shorter and cleaner.

File: deploy/templates/nstemplatetiers/test/ns_stage.yaml [hunk 13]
```diff
@@ -123,6 +65,20 @@ objects:
...
+    namespace: ${USERNAME}-stage
+  spec:
+    limits:
+    - type: "Container"
+      default:
+        cpu: 1000m
+        memory: 750Mi
+      defaultRequest:
+        cpu: 10m
+        memory: 64Mi
```
Comment (MatousJobanek) [inline deploy/templates/nstemplatetiers/test/ns_stage.yaml:81]: is this intentional?

File: deploy/templates/nstemplatetiers/test/ns_dev.yaml [hunk 14]
```diff
@@ -137,6 +79,20 @@ objects:
...
+    namespace: ${USERNAME}-dev
+  spec:
+    limits:
+    - type: "Container"
+      default:
+        cpu: 1000m
+        memory: 750Mi
+      defaultRequest:
+        cpu: 10m
+        memory: 64Mi
```
Comment (MatousJobanek) [inline deploy/templates/nstemplatetiers/test/ns_dev.yaml:95]: is this intentional?

File: controllers/space/space_controller_test.go (hunk 11 above)
Comment (xcoulon) [reply controllers/space/space_controller_test.go:1091]: good point, let me do that

File: controllers/space/space_controller_test.go (hunk 12 above)
Comment (xcoulon) [reply controllers/space/space_controller_test.go:1104]: ok, let me check

File: deploy/templates/nstemplatetiers/test/ns_stage.yaml (hunk 13 above)
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_stage.yaml:81]: I don't think so 🤨

File: deploy/templates/nstemplatetiers/test/ns_dev.yaml (hunk 14 above)
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_dev.yaml:95]: I don't think so 🤨

File: controllers/space/space_controller_test.go (hunk 11 above)
Comment (xcoulon) [reply controllers/space/space_controller_test.go]: done in de6d17a

File: controllers/space/space_controller_test.go (hunk 12 above)
Comment (xcoulon) [reply controllers/space/space_controller_test.go]: done in 9d4b599

File: deploy/templates/nstemplatetiers/test/ns_stage.yaml (hunk 13 above)
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_stage.yaml]: removed in f43166a

File: deploy/templates/nstemplatetiers/test/ns_dev.yaml (hunk 14 above)
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_dev.yaml]: removed in f43166a

File: pkg/templates/nstemplatetiers/nstemplatetier_generator.go [hunk 15]
```diff
@@ -297,11 +298,12 @@ func (t *tierGenerator) createTierTemplates() error {
 	// create the templates
 	for _, tierTmpls := range t.templatesByTier {
 		for _, tierTmpl := range tierTmpls.tierTemplates {
+			log.Info("creating TierTemplate", "tiertemplate", spew.Sdump(tierTmpl))
 			// using the "standard" client since we don't need to support updates on such resources, they should be immutable
 			if err := t.client.Create(context.TODO(), tierTmpl); err != nil && !apierrors.IsAlreadyExists(err) {
 				return errors.Wrapf(err, "unable to create the '%s' TierTemplate in namespace '%s'", tierTmpl.Name, tierTmpl.Namespace)
 			}
-			log.Info("TierTemplate resource created", "namespace", tierTmpl.Namespace, "name", tierTmpl.Name)
+			log.Info("TierTemplate resource created", "namespace", tierTmpl.Namespace, "name", tierTmpl.Name, "tier_name", tierTmpl.Spec.TierName, "revision", tierTmpl.Spec.Revision, "template_name", tierTmpl.Spec.Template.GetName())
```
Comment (ranakan19) [inline pkg/templates/nstemplatetiers/nstemplatetier_generator.go:306]: Is there any value in adding the revision to log? I don't see any dependency changes introduced in this PR regarding the revision, maybe I missed it. I don't have an objection to having the revision added to logs, but would like to understand how looking at revision can be helpful to anybody looking at the logs.

File: controllers/space/space_controller_test.go [hunk 16]
```diff
@@ -1270,49 +1425,70 @@ func requestFor(s *toolchainv1alpha1.Space) reconcile.Request {
...
-		// when
-		setSpec := space.NewNSTemplateSetSpec(nsTemplateTier)
-
-		// then
-		assert.Equal(t, newExpectedNsTemplateSetSpec(), setSpec)
-	})
+func TestNewNSTemplateSetSpec(t *testing.T) {
 
-	t.Run("when clusterResources template is NOT specified", func(t *testing.T) {
+	t.Run("full monty", func(t *testing.T) {
```
Comment (ranakan19) [inline controllers/space/space_controller_test.go:1430]: its not clear to me what is full monty, could you please clarify or consider changing the name of this test case?

**Review by ranakan19 — COMMENTED** [review]
Looks good, left a few comments

File: deploy/templates/nstemplatetiers/test/ns_dev.yaml [hunk 17]
```diff
@@ -13,64 +13,6 @@ objects:
...
-    name: user-edit
-    namespace: ${USERNAME}-dev
-  roleRef:
-    apiGroup: rbac.authorization.k8s.io
-    kind: ClusterRole
-    name: edit
-  subjects:
-  - kind: User
-    name: ${USERNAME}
-- apiVersion: v1
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_dev.yaml:60]: restored in 366c505

File: deploy/templates/nstemplatetiers/test/ns_stage.yaml [hunk 18]
```diff
@@ -13,64 +13,6 @@ objects:
...
-    namespace: ${USERNAME}-stage
-  roleRef:
-    apiGroup: rbac.authorization.k8s.io
-    kind: ClusterRole
-    name: edit
-  subjects:
-  - kind: User
-    name: ${USERNAME}
-- apiVersion: v1
-  kind: LimitRange
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/test/ns_stage.yaml:61]: restored in 366c505

File: config/crd/bases/toolchain.dev.openshift.com_tiertemplates.yaml [hunk 19]
```diff
@@ -141,6 +141,7 @@ spec:
                 required:
                 - objects
                 type: object
+                x-kubernetes-embedded-resource: true
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply config/crd/bases/toolchain.dev.openshift.com_tiertemplates.yaml:144]: we could, indeed. To be honest, I have not checked if there are other places where this could be needed. This annotation is used to retain the metadata of the embedded resources, that's all.

File: controllers/space/space_controller_test.go (hunk 16 above)
Comment (xcoulon) [reply controllers/space/space_controller_test.go]: right, I did not realize that when I started working on this test. I removed the subtest wrapper in ba329fa, since there was no need for it anyways.

File: pkg/templates/nstemplatetiers/nstemplatetier_generator.go (hunk 15 above)
Comment (xcoulon) [reply pkg/templates/nstemplatetiers/nstemplatetier_generator.go]: yeah, that's something I added while working on this PR, but it's not very valuable. Reverted in 7fd75c7

File: deploy/templates/nstemplatetiers/appstudio/spacerole_admin.yaml [hunk 20]
```diff
@@ -57,6 +69,24 @@ objects:
...
+  - apiGroups:
+    - results.tekton.dev
+    resources:
+    - results
+    - records
+    verbs:
+    - get
+    - list
+  - apiGroups:
+    - singapore.open-cluster-management.io
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply deploy/templates/nstemplatetiers/appstudio/spacerole_admin.yaml:81]: it was added in https://github.com/codeready-toolchain/host-operator/commit/f3e55e00953b1b13d34ab63876b210e73049e0af and I moved from `deploy/templates/nstemplatetiers/appstudio/ns_appstudio.yaml` into `deploy/templates/nstemplatetiers/appstudio/spacerole_admin.yaml`

**Review by rajivnathan — COMMENTED** [review]
Looks good overall! 👍  Just added a few minor comments

File: controllers/space/space_controller.go [hunk 21]
```diff
@@ -309,6 +324,34 @@ func NewNSTemplateSetSpec(tmplTier *toolchainv1alpha1.NSTemplateTier) toolchainv
...
+			// no need to add an entry in space roles if there is no associated user
+			if len(usernames) > 0 {
+				sort.Strings(usernames)
+				s.SpaceRoles = append(s.SpaceRoles, toolchainv1alpha1.NSTemplateSetSpaceRole{
+					TemplateRef: sr.TemplateRef,
+					Usernames:   usernames,
+				})
+			}
+		}
+		fmt.Printf("Space roles: %+v\n", s.SpaceRoles)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/space_controller.go:353]: 🤦‍♂️

File: controllers/space/space_controller_test.go [hunk 22]
```diff
@@ -1029,6 +1031,159 @@ func TestUpdateSpaceTier(t *testing.T) {
...
+
+		ctrl := newReconciler(hostClient, member1, member2)
+
+		// when
+		res, err := ctrl.Reconcile(context.TODO(), requestFor(s))
+
+		// then
+		require.NoError(t, err)
+		assert.True(t, res.Requeue) // expect a requeue since the NSTemplateSet was updated
+		// Space should be in "updating" state while the
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/space_controller_test.go:1083]: oh my

[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/space_controller_test.go:1083]: (responding with an incomplete comment as well 😂)

File: test/templates/nstemplatetiers/base/spacerole_admin.yaml [hunk 23]
```diff
@@ -9,7 +9,7 @@ objects:
   kind: RoleBinding
   metadata:
     namespace: ${NAMESPACE}
-    name: user-rbac-edit
+    name: ${USERNAME}-user-rbac-edit
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply test/templates/nstemplatetiers/base/spacerole_admin.yaml]: good point! done in cb8900f

File: test/nstemplatetier/nstemplatetier.go [hunk 24]
```diff
@@ -72,6 +80,14 @@ var CurrentBasicTemplates = toolchainv1alpha1.NSTemplateTierSpec{
 	ClusterResources: &toolchainv1alpha1.NSTemplateTierClusterResources{
 		TemplateRef: "basic-clusterresources-123456new",
 	},
+	SpaceRoles: map[string]toolchainv1alpha1.NSTemplateTierSpaceRole{
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply test/nstemplatetier/nstemplatetier.go:91]: done in cab6180

File: controllers/space/space_controller.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/space_controller.go]: removed in 289d2ea

File: controllers/space/space_controller_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply controllers/space/space_controller_test.go]: fixed in 7e52bbb

**Review by MatousJobanek — APPROVED** [review]
Looks good, thanks for addressing all my comments :pray: 
Please, before mering this PR:
1. Let's think about all possible corner-cases of this migration step that could affect the end-user when the PR is merged
2. Coordinate in a similar way as we did it previously so we can monitor the migration flow and possible errors.

**Review by rajivnathan — APPROVED** [review]
Thanks for addressing all my comments! 👍
